### PR TITLE
Созонов Илья. Вариант 26. Технология OMP. Линейная фильтрация изображений (блочное разбиение). Ядро Гаусса 3x3.

### PR DESCRIPTION
**Pull request by sozozzya**
**Стандартная свертка Гаусса (3×3):**

Ядро Гаусса 3x3 (нормированное) для σ=1 имеет вид: 

image

Каждому пикселю изображения ставится в соответствие новая величина, полученная как сумма произведений значений пикселей из окрестности 3×3 на коэффициенты ядра Гаусса.

Например, если центральный пиксель изображения **𝐼(𝑥, 𝑦)**, то его новое значение **𝐼′(𝑥, 𝑦)** будет: 

image

**Параллельная реализация:**

Для ускорения вычислений разбиваем изображение на блоки и обрабатываем их параллельно, используя OpenMP.

**1. Разбиение на блоки**
- Определяем размер блоков `block_size`.
- Двигаемся по изображению, обрабатывая его кусочками, а не по всей ширине/высоте сразу.

**2. Параллельная обработка**
- Используем `#pragma omp parallel for schedule(dynamic, 1)`, чтобы каждый поток обрабатывал по одной строке блоков.
- Внутри каждого блока применяем стандартную свертку Гаусса (3×3), но только к пикселям внутри блока.
- Учитываем границы блоков, чтобы избежать выхода за границы массива.

**3. Запись результата**
- После завершения вычислений все потоки записывают отфильтрованные данные в общий массив.

### CODE REVIEW

**Review by Ivan-Lysov — APPROVED** [review]
add a test for an empty image

File: tasks/omp/sozonov_i_image_filtering_block_partitioning/src/ops_omp.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+bool sozonov_i_image_filtering_block_partitioning_omp::TestTaskOpenMP::RunImpl() {
+  std::vector<double> kernel = {0.0625, 0.125, 0.0625, 0.125, 0.25, 0.125, 0.0625, 0.125, 0.0625};
+
+  int block_size = 32;
+
+  int num_blocks_x = (width_ + block_size - 1) / block_size;
+  int num_blocks_y = (height_ + block_size - 1) / block_size;
+
+#pragma omp parallel for schedule(dynamic)
```
Comment (ascannel) [inline tasks/omp/sozonov_i_image_filtering_block_partitioning/src/ops_omp.cpp:50]: I suggest you to specify chunk size for schedule

**Review by ascannel — CHANGES_REQUESTED** [review]

**Review by ascannel — APPROVED** [review]

File: tasks/omp/sozonov_i_image_filtering_block_partitioning/src/ops_omp.cpp (hunk 1 above)
Comment (sozozzya) [reply tasks/omp/sozonov_i_image_filtering_block_partitioning/src/ops_omp.cpp]: fixed, please check

File: tasks/omp/sozonov_i_image_filtering_block_partitioning/perf_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  std::vector<double> ans(width * height, 1);
+
+  for (int i = 0; i < width; ++i) {
+    ans[i] = 0;
+    ans[((height - 1) * width) + i] = 0;
+  }
+  for (int i = 1; i < height - 1; ++i) {
+    ans[i * width] = 0;
+    ans[(i * width) + width - 1] = 0;
+  }
```
Comment (allnes) [inline tasks/omp/sozonov_i_image_filtering_block_partitioning/perf_tests/main.cpp:28]: please move to common function for both test

Comment (sozozzya) [reply tasks/omp/sozonov_i_image_filtering_block_partitioning/perf_tests/main.cpp]: moved, please check

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]
